internal/graph/lookupsubjects.go

```diff
@@ -89,67 +176,256 @@ func subjectsForConcreteIds(subjectIds []string) map[string]*v1.FoundSubjects {
+				return cl.dispatchIndirectSubjectsForRelation(ctx, ci.withClonedLimits(), req, cstream, reader)
+			},
+			runIf: hasIndirectSubjects,
+		},
+	)
+}
+
+// lookupMatchingSubjectsForRelation finds all directly matching subjects on the request's relation, if applicable.
+func (cl *ConcurrentLookupSubjects) lookupMatchingSubjectsForRelation(
+	ctx context.Context,
+	ci cursorInformation,
+	req ValidatedLookupSubjectsRequest,
+	stream dispatch.LookupSubjectsStream,
+	validatedTS *namespace.ValidatedNamespaceTypeSystem,
+	reader datastore.Reader,
+) error {
+	// Check if the direct subject can be found on this relation and, if so, query for then.
+	directAllowed, err := validatedTS.IsAllowedDirectRelation(req.ResourceRelation.Relation, req.SubjectRelation.Namespace, req.SubjectRelation.Relation)
+	if err != nil {
+		return err
+	}
+
+	if directAllowed == namespace.DirectRelationNotValid {
+		return nil
+	}
+
+	var afterCursor options.Cursor
+	afterSubjectID, _ := ci.headSectionValue()
+
+	// If the cursor specifies the wildcard, then skip all further non-wildcard results.
+	if afterSubjectID == tuple.PublicWildcard {
+		return nil
+	}
+
+	if afterSubjectID != "" {
+		afterCursor = &core.RelationTuple{
+			// NOTE: since we fully specify the resource below, the resource should be ingored in this cursor.
+			ResourceAndRelation: &core.ObjectAndRelation{
+				Namespace: "",
+				ObjectId:  "",
+				Relation:  "",
+			},
+			Subject: &core.ObjectAndRelation{
+				Namespace: req.SubjectRelation.Namespace,
+				ObjectId:  afterSubjectID,
+				Relation:  req.SubjectRelation.Relation,
+			},
+		}
+	}
+
+	limit := ci.limits.currentLimit + 1 // +1 because there might be a matching wildcard too.
+	if !ci.limits.hasLimit {
+		limit = 0
+	}
+
+	foundSubjectsByResourceID := datasets.NewSubjectSetByResourceID()
+	if err := queryForDirectSubjects(ctx, req, datastore.SubjectsSelector{
+		OptionalSubjectType: req.SubjectRelation.Namespace,
+		RelationFilter:      datastore.SubjectRelationFilter{}.WithNonEllipsisRelation(req.SubjectRelation.Relation),
+	}, afterCursor, foundSubjectsByResourceID, reader, limit); err != nil {
+		return err
+	}
+
+	// Send the results to the stream.
+	if foundSubjectsByResourceID.IsEmpty() {
+		return nil
+	}
+	return publishSubjects(stream, ci, foundSubjectsByResourceID.AsMap())
+}
+
+// lookupWildcardSubjectForRelation finds the wildcard subject on the request's relation, if applicable.
+func (cl *ConcurrentLookupSubjects) lookupWildcardSubjectForRelation(
+	ctx context.Context,
+	ci cursorInformation,
+	req ValidatedLookupSubjectsRequest,
+	stream dispatch.LookupSubjectsStream,
+	validatedTS *namespace.ValidatedNamespaceTypeSystem,
+	reader datastore.Reader,
+) error {
+	// Check if a wildcard is possible and, if so, query directly for it without any cursoring. This is necessary because wildcards
+	// must *always* be returned, regardless of the cursor.
```

Comment:
Yes, because of intersections/exclusions: if you intersect a wildcard with a concrete, we need the wildcard returned to ensure the computation works as expected